File: leetcode/3sum-closet/3sum-closet.cpp
/**************************************
* @author dujiong
* @date 2016.11.29
* @version V0.1
**************************************/

class Solution {
public:
    int threeSumClosest(vector<int> &num, int target) {
        std::sort(num.begin(), num.end());
        int closet=65535;			
        int sum;
        int last = num[0];
        for(int i=0;i<num.size()-2;i++){
            if(i!=0 && num[i]==last) continue;
            int left=i+1;
            int right=num.size()-1;
            int ret;
            while(left<right){
                ret = num[i]+num[left]+num[right];
     			if(abs(ret-target)<closet){
                    closet = abs(ret-target);
                    sum = ret;
                }
                if(ret==target){
                    return sum;
                }else if(ret<target){
                    left++;
                }else{
                    right--;
                }
            }
            last = num[i];
        }
        return sum;
    }
};

File: leetcode-golang/3Sum Closet/3Sum Closet.go
import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	diff := math.MaxInt32
	res := 0
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		left, right := i+1, len(nums)-1
		for left < right {
			sum := nums[left] + nums[right]
			if nums[i]+sum == target {
				return 0
			} else if nums[i]+sum < target {
				left++
			} else {
				right--
			}
			if int(math.Abs(float64(sum+nums[i]-target))) < diff {
				diff = int(math.Abs(float64(sum + nums[i] - target)))
				res = sum + nums[i]
			}
		}
	}
	return res
}